File: solutions/javascript/robot-simulator/1/robot-simulator.js
export class InvalidInputError extends Error {
  constructor(message) {
    super();
    this.message = message || 'Invalid Input';
  }
}

export class Robot {
  constructor(){
    this.direction = "north";
    this.pos = [0,0]
  }
  get bearing() {
    return this.direction;
  }
  get coordinates() {
    return this.pos;
  }
  moveRobot() {
    if (this.direction==="north") this.pos[1]+=1
    else if (this.direction==="east") this.pos[0]+=1
    else if (this.direction==="south") this.pos[1]-=1
    else if (this.direction==="west") this.pos[0]-=1
  }
  place({ x, y, direction }) {
    if (!["north","south","east","west"].includes(direction)){
      throw new InvalidInputError()
    }
    this.pos = [x,y];
    this.direction = direction
  }
  static turnLeft = direction => {
    const turns = {north:"west",south:"east",east:"north",west:"south"}
    return turns[direction]
  }
  static turnRight = direction => Robot.turnLeft(Robot.turnLeft(Robot.turnLeft(direction)))
  evaluate(instructions) {
    for (const char of instructions){
      if (char==="R") this.direction = Robot.turnRight(this.direction)
      else if (char==="L") this.direction = Robot.turnLeft(this.direction)
      else this.moveRobot()
    }
  }
}

File: solutions/lua/robot-simulator/1/robot-simulator.lua
local Robot = {}
local RIGHT = {
  north = "east",
  east = "south",
  south = "west",
  west = "north"
}
local OFFSETS = {
  north = {x = 0, y = 1},
  east = {x = 1, y = 0},
  south = {x = 0, y = -1},
  west = {x = -1, y = 0}
}
function Robot:advance()
  local offset = OFFSETS[Robot.heading]
  Robot.x, Robot.y = Robot.x + offset.x, Robot.y + offset.y
end
function Robot:turn_right()
  Robot.heading = RIGHT[Robot.heading]
end
function Robot:move_single(instruction) 
  if instruction == "A" then
    Robot:advance()
  elseif instruction == "R" then
    Robot:turn_right()
  elseif instruction == "L" then
    Robot:turn_right()
    Robot:turn_right()
    Robot:turn_right()
  else
    error("invalid instruction")
  end
end

function Robot:move(instruction)
  for char in instruction:gmatch"." do
    Robot:move_single(char)
  end
end
return function(config)
  Robot.x, Robot.y, Robot.heading = config.x, config.y, config.heading
  return Robot
end
